src/dir_contents.ml

```diff
@@ -179,12 +179,36 @@ type modules =
     rev_map : Buildable.t Module.Name.Map.t
   }
 
+let empty_modules =
+  { libraries = String.Map.empty
+  ; executables = String.Map.empty
+  ; rev_map = Module.Name.Map.empty
+  }
 type t =
-  { text_files : String.Set.t
+  { kind : kind
+  ; dir : Path.t
+  ; text_files : String.Set.t
   ; modules : modules Lazy.t
   ; mlds : (Jbuild.Documentation.t * Path.t list) list Lazy.t
   }
 
+and kind =
+  | Standalone
+  | Group_root of t list Lazy.t
+  | Group_part of t
+
+let kind t = t.kind
+let dir t = t.dir
+
+let dirs t =
+  match t.kind with
+  | Standalone -> [t]
+  | Group_root (lazy l) -> t :: l
+  | Group_part t ->
+    match t.kind with
```

Comment:
The compiler seems to be doing fine, but I think adding `begin`/`end` around the nested `match` will make this easier to read. An alternative is to use deep pattern matching instead of the nested `match`, and move the `assert false` to the outer `match`.

Reply:
I used a deep pattern matching